CommonTools/RecoAlgos/src/PrimaryVertexAssignment.cc

```diff
@@ -4,50 +4,102 @@
 #include "TrackingTools/TransientTrack/interface/TransientTrackBuilder.h"
 #include "DataFormats/Math/interface/deltaR.h"
 #include "TrackingTools/IPTools/interface/IPTools.h"
-
+#include "FWCore/Utilities/interface/isFinite.h"
 
 
 std::pair<int,PrimaryVertexAssignment::Quality>
 PrimaryVertexAssignment::chargedHadronVertex( const reco::VertexCollection& vertices,
                                    const reco::TrackRef& trackRef,
                                    const reco::Track* track,
+                                   float time, 
+                                   float timeReso, // <0 if timing not available for this object
                                    const edm::View<reco::Candidate>& jets,
                                    const TransientTrackBuilder& builder) const {
 
-  int iVertex = -1;
-  size_t index=0;
   typedef reco::VertexCollection::const_iterator IV;
   typedef reco::Vertex::trackRef_iterator IT;
+
+  int iVertex = -1;
+  size_t index=0;
   float bestweight=0;
   for( auto const & vtx : vertices) {
       float w = vtx.trackWeight(trackRef);
       if (w > bestweight){
-          bestweight=w;
-          iVertex=index;
-        }        
+        bestweight=w;
+        iVertex=index;
+      }        
       index++;
   }
+  
+  bool useTime = useTiming_;
+  if (edm::isNotFinite(time) || timeReso<1e-6) {
+    useTime = false;
+    time = 0.;
+    timeReso = -1.;
+  }
+  
+  if (preferHighRanked_) {
+    for(IV iv=vertices.begin(); iv!=vertices.end(); ++iv) {
+        int ivtx = iv - vertices.begin();
+        if (iVertex == ivtx) return std::pair<int,PrimaryVertexAssignment::Quality>(ivtx,PrimaryVertexAssignment::UsedInFit);
+        
+        double dz = std::abs(track->dz(iv->position()));
+        double dt = std::abs(time-iv->t());
+        
+        bool useTimeVtx = useTime && iv->tError()>0.;
+        
+        if ((dz < maxDzForPrimaryAssignment_ or dz/track->dzError() < maxDzSigForPrimaryAssignment_ ) and (!useTimeVtx or dt/timeReso < maxDtSigForPrimaryAssignment_)) {
```

Comment:
why is the vertex error not a part of the significance computation here (it is in other places in this file)?

Reply:
the vertex supposedly has a time resolution at this point. 
Shouldn't it be used as well in the significance computation?

Reply:
The vertex time error is only checked if it's >0 (ie to check that the vertex has valid timing info.)

The numerical value is intentionally not used in order to avoid preferring to assign tracks to vertices with large time uncertainty (and the logic is coherent for z and t).

Reply:
(where the dz significance is used later in the code including the vertex error, it will not cause poor vertices to be favoured, because it is only checked against the closest vertex)

Reply:
the reason to add the vtx z error was to "regularize" the selection for tracks with very small dzError. IIRC, values down to 5 um are possible which can simply kill a good track with respect to still rather reasonable vertices.